messaging/src/main/java/org/axonframework/serialization/json/JacksonConverter.java

```diff
@@ -16,124 +16,131 @@
-        } catch (JsonProcessingException e) {
-            var errorMessage = """
-                Failed to convert between %s and %s: %s
-                """.formatted(sourceType.getSimpleName(), targetType.getSimpleName(), e.getMessage());
-            logger.error(errorMessage, e);
-            throw new ConversionException(errorMessage, e);
-        }
-    }
-
-    @SuppressWarnings("unchecked")
-    private <S, T> T performConversion(S input, Class<S> sourceType, Class<T> targetType)
-            throws JsonProcessingException {
-
-        // Handle serialization: event -> byte[] or String
-        if (canSerialize(targetType)) {
-            var json = objectMapper.writeValueAsString(input);
-            return switch (targetType.getName()) {
-                case "[B" -> (T) json.getBytes(StandardCharsets.UTF_8); // byte[]
-                case "java.lang.String" -> (T) json;
-                default -> throw new IllegalArgumentException("Unsupported target type: " + targetType);
-            };
-        }
-
-        // Handle deserialization: byte[] or String -> event
-        var jsonInput = switch (input) {
-            case byte[] bytes -> new String(bytes, StandardCharsets.UTF_8);
-            case String str -> str;
-            default -> throw new IllegalArgumentException("Unsupported source type: " + sourceType);
-        };
-
-        return objectMapper.readValue(jsonInput, targetType);
-    }
-
-    /**
-     * Custom exception for conversion failures.
-     */
-    public static final class ConversionException extends RuntimeException {
-
-        /**
-         * Constructs a new ConversionException with the specified detail message.
-         * @param message The detail message.
-         * @param cause The cause of the exception.
-         */
-        public ConversionException(String message, Throwable cause) {
-            super(message, cause);
+            JavaType targetJavaType = objectMapper.constructType(targetType);
+            if (converter.canConvert(sourceType, targetJavaType.getRawClass())) {
+                if (logger.isTraceEnabled()) {
+                    logger.trace(
+                            "Converter [{}] will do the entire conversion from source [{}] to target [{}] for [{}].",
+                            converter, sourceType, targetType, input
+                    );
+                }
+                //noinspection unchecked
+                return (T) converter.convert(input, targetJavaType.getRawClass());
+            } else if (converter.canConvert(sourceType, byte[].class)) {
+                if (logger.isTraceEnabled()) {
+                    logger.trace("Converts input [{}] to byte[] before reading it into [{}].", input, targetJavaType);
+                }
+                return objectMapper.readValue(converter.convert(input, byte[].class), targetJavaType);
+            } else if (converter.canConvert(targetJavaType.getRawClass(), byte[].class)) {
```

Comment:
This branch does not make sense to me :thinking: 
If the target type can be converted to bytes, then we write the input to bytes? 
That doesn't seem the right condition to me, should this be:

```converter.canConvert(byte[].class, targetJavaType.getRawClass())```?

Reply:
Aaah, you're right, this should be flipped. Will do in #3558. Thanks!